File: Dynamic_Programming/2008. Maximum Earnings From Taxi.cpp
/*
2008. Maximum Earnings From Taxi
*/

// TC:O(m+n+nlgn)
class Solution {
    using LL = long long;
public:
    long long maxTaxiEarnings(int n, vector<vector<int>>& rides) {
        vector<LL> dp(n+1);
        auto cmp = [](const vector<int>&a, const vector<int>& b){
            return a[1] < b[1];
        };
        sort(rides.begin(), rides.end(), cmp);
        int idx = 0;
        for(int i = 1; i <= n; i++){
            dp[i] = dp[i-1];
            while(idx < rides.size() && i == rides[idx][1]){
                LL start = rides[idx][0], end = rides[idx][1], tip = rides[idx][2];
                dp[i] = max(dp[i], dp[start] + end-start+tip); 
                idx++;
            }
        }

        return dp[n];
    }
};

/*
這題問說給你一堆客人上車與下車的點 還有會給小費
問你你一路開到底可以賺到的最多錢是多少 end-start + tip
可以放人下車與讓人上車在同一點
dp[i]定義成走到i點能獲取的最大收益
用二分去找出跟他起點一樣或者小於的點的最大收益
 x x x 
    x x x   
x x x x x x x 
              x x x  
                  x x x x
                x x x x x 
*/

File: Dynamic_Programming/2008. Maximum Earnings From Taxi.py
# 不排序: TC:O(n) SC:O(n) 排序 + 二分 : TC:O(mlgm) SC:O(m) m = 3 * 1e3,n = 1e5
class Solution:
    def maxTaxiEarnings(self, n: int, rides: List[List[int]]) -> int:
        # 當數據很大的話需排序來做 
        rides.sort(key=lambda x:x[1]) # sort by ending point
        drop_offs = [ride[1] for ride in rides] #用來二分
        dp = [0] * (len(rides) + 1)
        for i in range(1, len(rides) + 1):
            start, end, tip = rides[i-1]
            j = bisect_right(drop_offs, start)
            dp[i] = max(dp[i-1], dp[j] + end - start + tip)
        return dp[-1]
        '''
        f = [0] * (n + 1) # from 1 to n
        roads = [[] for _ in range(n + 1)]
        for st, end, tip in rides:
            roads[end].append((st, tip))
        for r, g in enumerate(roads): # 目前的終點 以及當下的點是否可以載客到目的地
            f[r] = f[r-1] # 在數據範圍1e5內的可以用這個避免排序or二分
            for st, tip in g:
                if f[st] + (r - st + tip) > f[r]:
                    f[r] = f[st] + (r  - st + tip)
        return f[n]
        '''
